src/btree/bt_curnext.c

```diff
@@ -427,11 +510,28 @@ __cursor_row_next(WT_CURSOR_BTREE *cbt, bool newpage, bool restart, size_t *skip
              * If the cursor has prefix search configured we can early exit here if the key that we
              * are visiting is after our prefix.
              */
-            if (prefix_search && __wt_prefix_match(prefix, key) < 0) {
-                *prefix_key_out_of_bounds = true;
+            if (prefix_search && __wt_prefix_match(prefix, &cbt->iface.key) < 0) {
+                *key_out_of_bounds = true;
                 WT_STAT_CONN_DATA_INCR(session, cursor_search_near_prefix_fast_paths);
                 return (WT_NOTFOUND);
             }
+
+            /*
+             * If an upper bound has been set ensure that the key is within the range, otherwise
+             * early exit.
+             */
+            if (F_ISSET(cursor, WT_CURSTD_BOUND_UPPER)) {
+                WT_ASSERT(session, WT_DATA_IN_ITEM(&cbt->iface.upper_bound));
+                WT_RET(
+                  __wt_row_compare_bounds(session, &cbt->iface, btree->collator, true, &out_range));
+                /* Check that the key is within the range if bounds have been set. */
+                if (out_range) {
```

Comment:
Instead of redefining out_range just pass key_out_of_bounds to the compare bounds function and test that in the if statements.